miden-crypto/src/merkle/smt/partial.rs

```diff
@@ -314,38 +332,188 @@ impl PartialSmt {
+    /// if not stored.
+    fn get_inner_node_or_empty(&self, index: NodeIndex) -> InnerNode {
+        self.get_inner_node(index)
+            .unwrap_or_else(|| EmptySubtreeRoots::get_inner_node(SMT_DEPTH, index.depth()))
+    }
+
+    /// Inserts an inner node at the specified index.
+    fn insert_inner_node(&mut self, index: NodeIndex, inner_node: InnerNode) {
+        if inner_node == EmptySubtreeRoots::get_inner_node(SMT_DEPTH, index.depth()) {
+            self.inner_nodes.remove(&index);
+        } else {
+            self.inner_nodes.insert(index, inner_node);
+        }
+    }
+
+    /// Returns the merkle path for a key by walking up the tree from the leaf.
+    fn get_path(&self, key: &Word) -> SparseMerklePath {
+        let index = NodeIndex::from(Self::key_to_leaf_index(key));
+
+        // Use proof_indices to get sibling indices from leaf to root,
+        // and get each sibling's hash
+        SparseMerklePath::from_sized_iter(index.proof_indices().map(|idx| self.get_node_hash(idx)))
+            .expect("path should be valid since it's from a valid SMT")
+    }
+
+    /// Get the hash of a node at an arbitrary index, including the root or leaf hashes.
+    ///
+    /// The root index simply returns the root. Other hashes are retrieved by looking at
+    /// the parent inner node and returning the respective child hash.
+    fn get_node_hash(&self, index: NodeIndex) -> Word {
+        if index.is_root() {
+            return self.root;
+        }
+
+        let InnerNode { left, right } = self.get_inner_node_or_empty(index.parent());
+
+        if index.is_value_odd() { right } else { left }
+    }
+
+    /// Recomputes all inner nodes from a leaf up to the root after a leaf value change.
+    fn recompute_nodes_from_leaf_to_root(
+        &mut self,
+        leaf_index: LeafIndex<SMT_DEPTH>,
+        leaf_hash: Word,
+    ) {
+        use crate::hash::rpo::Rpo256;
+
+        let mut index: NodeIndex = leaf_index.into();
+        let mut node_hash = leaf_hash;
+
+        for _ in (0..index.depth()).rev() {
+            let is_right = index.is_value_odd();
+            index.move_up();
+            let InnerNode { left, right } = self.get_inner_node_or_empty(index);
+            let (left, right) = if is_right {
+                (left, node_hash)
+            } else {
+                (node_hash, right)
+            };
+            node_hash = Rpo256::merge(&[left, right]);
+
+            // insert_inner_node handles removing empty subtree roots
+            self.insert_inner_node(index, InnerNode { left, right });
+        }
+        self.root = node_hash;
+    }
+
+    /// Validates the internal structure during deserialization.
+    ///
+    /// Checks that:
+    /// - Each inner node's hash is consistent with its parent.
+    /// - Each leaf's hash is consistent with its parent inner node's left/right child.
+    fn validate(&self) -> Result<(), DeserializationError> {
```

Comment:
Should we add a small test covering an inner node hash mismatch and a leaf hash mismatch to make sure validation is working as expected?